### PR TITLE
[CUDA] Add support for CUDA-12.3 and sm_90a

### PR DESCRIPTION
**Pull request by Artem-B**

### CODE REVIEW

**Comment by Artem-B** [comment]
@ezhulenev FYI.

File: llvm/lib/Target/NVPTX/NVPTXSubtarget.h [hunk 1]
```diff
@@ -80,8 +85,10 @@ class NVPTXSubtarget : public NVPTXGenSubtargetInfo {
   bool allowFP16Math() const;
   bool hasMaskOperator() const { return PTXVersion >= 71; }
   bool hasNoReturn() const { return SmVersion >= 30 && PTXVersion >= 64; }
-  unsigned int getSmVersion() const { return SmVersion; }
+  unsigned int getSmVersion() const { return FullSmVersion / 10; }
+  unsigned int getFullSmVersion() const { return FullSmVersion; }
   std::string getTargetName() const { return TargetName; }
+  bool isSm90a() const { return getFullSmVersion() == 901; }
```
Comment (jhuber6) [inline llvm/lib/Target/NVPTX/NVPTXSubtarget.h:91]: Could we expose this more like `getSmVersion` and `getSmFeature`? Has CUDA even documented how they intend to further build on this?

Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXSubtarget.h:91]: According to [CUDA docs](docs.nvidia.com/cuda/parallel-thread-execution/index.html?highlight=sm_90a#ptx-module-directives-target)

> Target architectures with suffix “a”, such as sm_90a, include architecture-accelerated features that are supported on the specified architecture only, hence such targets do not follow the onion layer model. Therefore, PTX code generated for such targets cannot be run on later generation devices. Architecture-accelerated features can only be used with targets that support these features.

It's not clear where they are going with this approach.

I can make it a more generic `int hasAAFeatures() { return FullSmVersion % 10;  }` if that's what you're looking for.

Comment (jhuber6) [reply llvm/lib/Target/NVPTX/NVPTXSubtarget.h:91]: Yeah, I was thinking that the internal representation would just be what "FullSMVersion" is now, but `getSMVersion` would return `/ 10` and `getFeatures` or something would be `% 10`.

Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXSubtarget.h]: Done.

**Review by jhuber6 — APPROVED** [review]

**Comment by Artem-B** [comment]
Tested the changes with cuda test-suite, with cuda-12.1 and 12.3 targeting `sm_{60,70,80,90,90a}`.

**Pull request merged**